packages/framework/framework/src/telemetry/index.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  Tracer as OTTracer,
+  Span,
+  trace,
+  context,
+  propagation,
+} from "@opentelemetry/api"
+
+/**
+ * Tracer creates an instrumentation scope within the application
+ * code. For example: You might create a tracer for the API
+ * requests, another one for the modules, one for workflows
+ * and so on.
+ *
+ * There is no need to create a Tracer instance per HTTP
+ * call.
+ */
+export class Tracer {
+  /**
+   * Reference to the underlying OpenTelemetry tracer
+   */
+  #otTracer: OTTracer
+  constructor(public name: string, public version?: string) {
+    this.#otTracer = trace.getTracer(name, version)
+  }
+
+  /**
+   * Returns the underlying tracer from open telemetry that
+   * could be used directly for certain advanced use-cases
+   */
+  getOTTracer() {
+    return this.#otTracer
+  }
+
+  /**
+   * Trace a function call. Using this method will create a
+   * child scope for the invocations within the callback.
+   */
+  trace<F extends (span: Span) => unknown>(
```

Comment:
Not actionable, just talking out loud - having the tracer take a callback means that all functions will have additional indentation, which makes readability a bit worse.

Now the good thing is, you will rarely need to trace manually, so it's probably not an issue. Also, I can't think of a better interface for it that won't introduce another confusion (eg. having to manually do `span.end` at the end of a function).

Reply:
The issue comes directly from OpenTelemetry SDK. In order to create parent/child relationship between span, you have to use `tracer.startActiveSpan` method, which accepts a callback.

There is also `tracer.startSpan` method, but there isn't any way to then assign a parent span to it. For example:

```ts
const parentSpan = tracer.startSpan('parent')

// There is no way for me to tell the child span to be part of the parent span
const childSpan = tracer.startSpan('child')
```

Whereas in the following example:

```ts
tracer.startActiveSpan('parent', (parentSpan) => {
  tracer.startActiveSpan('child', (childSpan) => {
    // childSpan is automatically child of parentSpan
  })
})
```

Reply:
Yeah completely understandable, I was just talking out loud, I know there isn't a lot we can do about it :)